### DISCUSSION TITLE
Experience with Tkinter for the UI MVP

### DISCUSSION THREAD

**Discussion by Coffee2Bits** [DevOps-MBSE/AaC#131] [General]
Per JD's [comment](https://github.com/jondavid-black/AaC/pull/111#pullrequestreview-820168367), I'm posting my thoughts about working with Tkinter for the UI.

My first impression: I didn't enjoy working with Tkinter. 

CONS:
* Long-lived/Old
  * Documentation/Solution changes
  * Several iterations of 'improved' packages, making it less coherent overall as a framework
* Inconsistency in stylizing Tkinter.tk components and Tkinter.ttk components (and it's recommended to use .ttk modules over .tk modules)
* Lacking a component framework -- making users more prone to duplicate code
* Limited widget set (limits what we can do out of the box)
* Lack of development tools (such as browser-integrated web development tools)
* Looks like a very old/low-effort UI

PROS:
* Easily accessible
* Easily integrated into the application
* Lots of documentation/Q&A on the topic
* Lots of simple widget tutorials

Tkinter is pretty limited in what it provides, we'll need to implement custom widgets just for something basic like a scroll window, and a lot of the difficulty I experienced developing the MVP is because I didn't have the same level of UI development tools or much of the same flexibility as when I'm building a web-based UI. 

I would have killed for CSS styling and Chrome DevTools on this effort, but I'm pretty biased towards web-based UIs even for applications like this. Web-based UIs work well enough for the likes of Slack, Spotify, Discord, and even [Skype](https://www.windowscentral.com/microsoft-swaps-skype-over-electron-windows-10). 

That being said, I think we should strongly consider an alternative to Tkinter. These are the top viable alternatives that I like so far:
*  [PyWebView](https://github.com/r0x0r/pywebview)*
* [Pyforms Web](https://github.com/UmSenhorQualquer/pyforms-web)
* [Kivy](https://kivy.org/doc/stable/guide/lang.html)

* Pywebivew looks like it allows us to use browser-based dev tools when [debugging the UI ](https://pywebview.flowrl.com/guide/debugging.html#debugging)

**Comment by jondavid-black** [comment]
I added a Q&A to capture everyone's inputs for a UI Framework trade study ( #138 ).  Please review and contribute your thoughts to that post.  Based on your feedback here and what it required to get the MVP working, I'm in agreement a change early is in our best interests.
